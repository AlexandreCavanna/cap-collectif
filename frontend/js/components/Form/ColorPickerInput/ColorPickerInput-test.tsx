/* eslint-env jest */
import * as React from 'react'
import { shallow } from 'enzyme'
import ColorPickerInput from './ColorPickerInput'

const props = {
  onChange: jest.fn(),
  value: '#ABCDEF',
}
describe('<ColorPickerInput />', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<ColorPickerInput {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
