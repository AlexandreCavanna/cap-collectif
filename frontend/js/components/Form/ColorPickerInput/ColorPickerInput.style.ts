import styled, { css } from 'styled-components'
import { MAIN_BORDER_RADIUS } from '~/utils/styles/variables'
import colors from '~/utils/colors'

export const ColorPickerInputContainer = styled.div.attrs({
  className: 'color-picker-input-container',
})<{
  hasValue: boolean
}>`
  position: relative;

  .chrome-picker {
    position: absolute;
    z-index: 1;
  }

  ${props =>
    props.hasValue &&
    css`
      .form-control {
        padding-left: 30px;
      }
    `};
`
export const PreviewColor = styled.div<{
  color: string
}>`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  left: 10px;
  background-color: ${props => props.color};
  border: 1px solid ${colors.borderColor};
  ${MAIN_BORDER_RADIUS};
  width: 15px;
  height: 15px;
`
