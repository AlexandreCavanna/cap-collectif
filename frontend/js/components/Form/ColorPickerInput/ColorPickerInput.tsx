import * as React from 'react'
import { FormControl } from 'react-bootstrap'
import ChromePicker from 'react-color/lib/Chrome'
import useClickAway from '~/utils/hooks/useClickAway'
import { ColorPickerInputContainer, PreviewColor } from './ColorPickerInput.style'

const hexToRgb = hex => {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex)
  return result
    ? {
        r: parseInt(result[1], 16),
        g: parseInt(result[2], 16),
        b: parseInt(result[3], 16),
      }
    : null
}

type Color = {
  hex: string
  hsl: {
    h: number
    s: number
    l: number
    a: number
  }
  oldHue: number
  rgb: {
    r: number
    g: number
    b: number
    a: number
  }
  source: string
}
type Props = {
  onChange: (value: string) => void
  value?: string
  label?: string
  disabled?: boolean
  placeholder?: string
  className?: string
  getOpacity?: (opacity: number) => void
  opacity?: number | null | undefined
}

const ColorPickerInput = ({
  onChange,
  value = '',
  placeholder,
  className,
  label,
  disabled,
  getOpacity,
  opacity,
}: Props) => {
  const [showPicker, setShowPicker] = React.useState<boolean>(false)
  const pickerReference = React.useRef(null)
  useClickAway(
    pickerReference,
    () => {
      if (showPicker) setShowPicker(false)
    },
    [],
  )

  const handleChange = (color: Color) => {
    const colorOpacity = color.rgb.a
    if (getOpacity && colorOpacity > 0 && colorOpacity < 100) getOpacity(colorOpacity * 100)
    onChange(color.hex)
  }

  const getValue = () => {
    if (opacity) {
      return { ...hexToRgb(value), a: opacity / 100 }
    }

    return value
  }

  return (
    <ColorPickerInputContainer ref={pickerReference} hasValue={!!value} className={className}>
      {value && <PreviewColor color={value} />}

      <FormControl
        type="text"
        label={label}
        disabled={disabled}
        onChange={onChange}
        value={value}
        placeholder={placeholder}
        onFocus={() => setShowPicker(true)}
      />

      {showPicker && <ChromePicker color={getValue()} onChangeComplete={handleChange} disableAlpha={false} />}
    </ColorPickerInputContainer>
  )
}

export default ColorPickerInput
