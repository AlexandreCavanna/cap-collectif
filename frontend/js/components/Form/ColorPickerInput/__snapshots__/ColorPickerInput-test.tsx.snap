// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPickerInput /> renders correctly 1`] = `
<ColorPickerInputstyle__ColorPickerInputContainer
  hasValue={true}
>
  <ColorPickerInputstyle__PreviewColor
    color="#ABCDEF"
  />
  <FormControl
    bsClass="form-control"
    componentClass="input"
    onChange={[MockFunction]}
    onFocus={[Function]}
    type="text"
    value="#ABCDEF"
  />
</ColorPickerInputstyle__ColorPickerInputContainer>
`;
